File: leetcode/23.cpp
#include <vector>
#include <queue>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  struct Node {
    int val;
    ListNode *ptr;
    Node(int x, ListNode *p) : val(x), ptr(p) {}
    bool operator<(const Node &n) const { return val > n.val; }
  };

  ListNode *mergeKLists(vector<ListNode *> &lists) {
    priority_queue<Node> cache;
    for (ListNode *node : lists) {
      if (node != nullptr) {
        cache.emplace(Node(node->val, node));
      }
    }
    ListNode sentry;
    ListNode *cur = &sentry;
    while (!cache.empty()) {
      const Node tmp = cache.top();
      cache.pop();
      cur->next = tmp.ptr;
      cur = cur->next;
      if (tmp.ptr->next != nullptr) {
        cache.push({tmp.ptr->next->val, tmp.ptr->next});
      }
    }
    return sentry.next;
  }
};

File: leetcode/23.py
import queue


# Definition for singly-linked list.
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Item(object):

    def __init__(self, node):
        self.node = node

    def __lt__(self, other):
        return self.node.val < other.node.val


class Solution:

    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        cache = queue.PriorityQueue(len(lists))
        sentry = ListNode()
        head = sentry

        for item in lists:
            if item is not None:
                cache.put(Item(item))

        while not cache.empty():
            item = cache.get()
            head.next = item.node
            head = head.next
            if head.next is not None:
                cache.put(Item(head.next))

        return sentry.next
